cluster-provision/gocli/cmd/provision_manager.go

```diff
@@ -0,0 +1,417 @@
+package cmd
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	fsys "kubevirt.io/kubevirtci/cluster-provision/gocli/cmd/filesystem"
+)
+
+const (
+	// Git status
+	FILE_ADDED   = "A"
+	FILE_DELETED = "D"
+	FILE_RENAMED = "R"
+
+	TARGET_NONE = "none"
+)
+
+type parameters struct {
+	tag       string
+	rulesFile string
+	debug     bool
+}
+
+type OutputSplitter struct{}
+
+var fileSystem fsys.FileSystem = fsys.RealFileSystem{}
+
+func SetFileSystem(fs fsys.FileSystem) {
+	if fs == nil {
+		fileSystem = fsys.RealFileSystem{}
+	} else {
+		fileSystem = fs
+	}
+}
+
+// NewProvisionManagerCommand determines which providers should be rebuilt
+func NewProvisionManagerCommand() *cobra.Command {
+	provision := &cobra.Command{
+		Use:   "pman",
+		Short: "provision manager determines which providers should be rebuilt",
+		RunE:  provisionManager,
+		Args:  cobra.ExactArgs(0),
+	}
+	provision.Flags().String("tag", "", "kubevirtci tag to compare to, default: fetch latest")
+	provision.Flags().String("rules", "hack/pman/rules.txt", "rules file, default: hack/pman/rules.txt")
+	provision.Flags().Bool("debug", false, "run in debug mode, default: false")
+
+	return provision
+}
+
+func provisionManager(cmd *cobra.Command, arguments []string) error {
+	params, err := parseArguments(cmd)
+	if err != nil {
+		return err
+	}
+
+	configLogger(params.debug)
+
+	// Sleep to let logrus flush its buffer, in order to avoid race between logrus and printing of the return value
+	defer func() {
+		time.Sleep(200 * time.Millisecond)
+	}()
+
+	if len(params.tag) == 0 {
+		params.tag, err = getKubevirtciTag()
+		if err != nil {
+			return err
+		}
+	}
+
+	printSection("Parameters")
+	logrus.Debug("Tag: ", params.tag)
+
+	targets, err := getTargets("cluster-provision/k8s/*")
+	if err != nil {
+		return err
+	}
+
+	rulesDB, err := buildRulesDBfromFile(params.rulesFile, targets)
+	if err != nil {
+		return err
+	}
+
+	targetToRebuild, err := processGitNameStatusChanges(rulesDB, targets, params.tag)
+	if err != nil {
+		return err
+	}
+
+	j, err := json.Marshal(targetToRebuild)
+	if err != nil {
+		return err
+	}
+
+	printSection("Result")
+	fmt.Println(string(j))
+
+	return nil
+}
+
+func parseArguments(cmd *cobra.Command) (parameters, error) {
+	params := parameters{}
+	var err error
+
+	params.debug, err = cmd.Flags().GetBool("debug")
+	if err != nil {
+		return parameters{}, err
```

Comment:
About `nil` (b), using an empty struct because:
it is a struct, so i can't return nil
`cmd/provision_manager.go:123:10: cannot use nil as parameters value in return statement`
unless we make it a pointer, but i think it is overkill

Reply:
about (a), Cobra already prints an informative error
for example use `--tag` without value gives `flag needs an argument: --tag`